File: 360. Sort Transformed Array/sortTransformedArray.py
from typing import List

class Solution:
  def calc(self, x: int, a: int, b: int, c: int) -> int:
    return a * x ** 2 + b * x + c

  def sortTransformedArray(self, nums: List[int], a: int, b: int, c: int) -> List[int]:
    l = len(nums)
    res = [0] * l
    left, right = 0, l - 1
    index = 0 if a <= 0 else l - 1

    caller = lambda x, y: x >= y if a >= 0 else x <= y

    while left <= right:
      leftVal = self.calc(nums[left], a, b, c)
      rightVal = self.calc(nums[right], a, b, c)

      if caller(leftVal, rightVal):
        res[index] = leftVal
        left += 1
      else:
        res[index] = rightVal
        right -= 1

      index += (1 if a <= 0 else -1)

    return res
    

nums = [-4,-2,2,4]
a = 1
b = 3
c = 5

solution = Solution()
print('sortTransformedArray: ', solution.sortTransformedArray(nums, a, b, c))

File: 360. Sort Transformed Array/sortTransformedArray.js
const calc = (x, a, b, c) => (a * x ** 2 + b * x + c);

const sortTransformedArray = (nums, a, b, c) => {
  const l = nums.length;
  const res = Array(l).fill(0);
  let left = 0;
  let right = l - 1;
  let index = a <= 0 ? 0 : l - 1;
  const compare = (x, y) => (a >= 0 ? x >= y : x <= y);

  while (left <= right) {
    let leftVal = calc(nums[left], a, b, c);
    let rightVal = calc(nums[right], a, b, c);

    if (compare(leftVal, rightVal)) {
      res[index] = leftVal;
      left++;
    } else {
      res[index] = rightVal;
      right--;
    }

    index += (a <= 0 ? 1 : -1);
  }

  return res;
};

const nums = [-4, -2, 2, 4];
const a = 1;
const b = 3;
const c = 5;
console.log('sortTransformedArray: ', sortTransformedArray(nums, a, b, c));
